allow fetchpatch everywhere, without restrictions

amjoseph-nixpkgs opened the pull request:
###### Description of changes

This PR eliminates all restrictions on the use of `fetchpatch`, and makes a point of using it in the earliest part of the `stdenv` bootstrap (i.e. `binutils`) to demonstrate the functionality and serve as a regression test going forward.

If this merges it can be followed by a treewide to delete all of the (untested and untestable) "cannot use fetchpatch" comments.

See also https://github.com/NixOS/nix/pull/7052; if you are running Nix >=2.12 you can revert the last two commits in this series.

###### Things done

- Built (`stdenv`) on platform(s)
  - [x] x86_64-linux
  - [x] powerpc64le-linux
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
patchutils: add forStdenvBootstrap flag

amjoseph-nixpkgs commented:
Ah, I see; there's no way to use `builtin:fetchurl` as part of an FOD unless it is the entire FOD.

Commit:
nix-fetchurl.nix: copy from nix repository

Commit:
nix-fetchurl.nix: add support for __impure

Commit:
fetchurl/boot.nix: accept postFetch parameter

Commit:
fetchurl/boot.nix: allow empty/fake hash

Commit:
stdenv/linux: use fetchpatch in stage0

Commit:
darwin/linux: use fetchpatch in stage0

Commit:
binutils: add demo and regression test for stage0-fetchpatch

Commit:
fetchurl: drop vendored fetchurl.nix

This commit assumes that https://github.com/NixOS/nix/pull/7052 merges.  If it
has not yet merged and you did not cherry-pick it into your copy of Nix, revert
this commit before testing.

SuperSandro2000 commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -188,6 +190,17 @@ in
+        inherit lib;
+        buildPackages.patchutils_0_3_3 = prevStage.patchutils_0_3_3.override {
+          forStdenvBootstrap = true;
+          # the parts of patchutils used by fetchpatch do not
+          # require perl, so we could eliminate this by patching
+          # patchutils' build scripts...
+          perl = bootstrapTools;
```
I am not sure if bootstrapTools contains anything related to perl. If not then I think it would be cleaner to exclude it like makeWrapper.

amjoseph-nixpkgs replied:
The `bootstrapTools` contain a copy of perl.  Mainly because autoconf/automake need it.

I would love to eliminate `perl` from the `bootstrapTools`.  I tried a while back and while it is probably possible it is a very major undertaking.  If it ever were eliminated we could (as my comment explains) apply a small patch to `patchutils` to drop all the parts of it that need perl (`fetchpatch` doesn't use those parts).

SuperSandro2000 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -68,11 +69,26 @@ stdenv.mkDerivation {
+      url = "https://sourceware.org/git/?p=binutils-gdb.git;a=patch;h=bef9ef8ca0f941d743c77cc55b5fe7985990b2a7";
+      revert = true;
+      excludes = [ "ChangeLog" ];
+      #hash = "sha256-VQyyPmY1P9NT7PYwu0vqocWmQGSyt45tADe6ecTc8fk=";
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -68,11 +69,26 @@ stdenv.mkDerivation {
+      #hash = "sha256-VQyyPmY1P9NT7PYwu0vqocWmQGSyt45tADe6ecTc8fk=";
+      hash = "sha256-FcTkFK9FfbgDR6iS+H6w7KRpPo/mFJ+0u7v5EvIlecQ=";
+    })
+    #./0001-Revert-libtool.m4-fix-nm-BSD-flag-detection.patch
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -68,11 +69,26 @@ stdenv.mkDerivation {
+    # this comment to that patch.  Having a `fetchpatch` in binutils
+    # serves as a regression test to ensure that there are no
+    # dependency cycles preventing it from being used in even the
+    # earliest stages of stdenv bootstrapping.
```
```suggestion
    # earliest stages of stdenv bootstrapping.
    #./0001-Revert-libtool.m4-fix-nm-BSD-flag-detection.patch
```

SuperSandro2000 commented on pkgs/build-support/fetchurl/boot.nix:
```diff
@@ -23,3 +28,21 @@ import <nix/fetchurl.nix> {
+  srcs = [ ];
+  unpackPhase = ''
+    runHook preUnpack
+    cp '${fetchurl (args // { postFetch=null; hash=""; sha256=""; __impure=true; })}' $out
```
```suggestion
    cp '${fetchurl (args // { postFetch = null; hash = ""; sha256 = ""; __impure=true; })}' $out
```

SuperSandro2000 reviewed (commented):
I think we must drop https://github.com/NixOS/nixpkgs/pull/188587/commits/30a099f3628b711b734ef9f18cab264061530b74 until the minimal nix version contains that fix :/

In addition to that can you do a treewide change to remove the warnings that fetchpatch cannot be used?

Commit:
Update pkgs/development/tools/misc/binutils/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Update pkgs/development/tools/misc/binutils/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Update pkgs/development/tools/misc/binutils/default.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Update pkgs/build-support/fetchurl/boot.nix

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
Revert "fetchurl: drop vendored fetchurl.nix"

This reverts commit 30a099f3628b711b734ef9f18cab264061530b74.

amjoseph-nixpkgs commented:
> I think we must drop [30a099f](https://github.com/NixOS/nixpkgs/commit/30a099f3628b711b734ef9f18cab264061530b74) until the minimal nix version contains that fix :/

I have reverted 30a099f.

https://github.com/NixOS/nix/pull/7052 just merged, so "minimal nix version" for 30a099f is now known to be "2.12".

Keep in mind that this PR still needs `impure-derivations` which is a Nix 2.8+ feature, so "minimal nix version" for the PR as a whole cannot go any lower than that -- unless there's some trick for adding a `postFetch` to `<nix/fetchurl.nix>` in older versions of Nix.


> In addition to that can you do a treewide change to remove the warnings that fetchpatch cannot be used?

Yes, once this PR is deemed mergeable I would be happy to do that (and squash these commits too).  But a treewide like that will accrue tons of merge conflicts if is is left sitting unmerged, so I'd rather wait until "the last minute" to push that commit into this PR.

Commit:
fetchurl/boot.nix: signal using `impure` rather than `outputHash=""`

The method for signaling that `__impure` is desired went through a few minor
revisions between this PR being filed and https://github.com/NixOS/nix/pull/7052
being merged.  This commit accounts for those changes so the "pre-Nix-2.12
emulation" uses the same interface that Nix 2.12 will use.

See also: https://github.com/NixOS/nix/pull/7052#discussion_r972785014

SuperSandro2000 commented:
@ofborg eval

amjoseph-nixpkgs commented:
> @ofborg eval

@SuperSandro2000, please see the giant text at the very top of the PR.

If there were some way I could add `<blink>` tags to it, I would.

amjoseph-nixpkgs commented:
I guess I should draftify this until/if impure-derivations is stabilized.

That may be years in the future, when this patch no longer applies, but I'm cool with that.  I just wanted to prove that this was possible.

amjoseph-nixpkgs commented:
This needs `ca-derivations`, and it looks like we're still a very long way from being able to use those.
